initial proposal for site-allowed-contracts (#1)
chainpatrol/proposals: proposals/site-allowed-contracts.md

# Website well-known Contracts [Draft]

### Draft

<aside>
🚧 This document is a work in progress, we are gathering feedback and suggestions to arrive at a complete specification.

</aside>

### Inspiration

Inspired by Content Security Policy (CSP) [https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP)

# Problem Definition

Currently there is no standard way to tell what contracts a website is expected to call. This means that a hacked site can request a signature for any contract, it can steal your tokens, cause you to sign a multisig compromising an organization, or otherwise damage the user.

> Another problem is that any site can call a smart contract, and this means that any site is part of the attack surface. A hack on an ecommerce store, a blog, or really any web page can lead to a legitimate site draining a user's wallet.

Reviewer: A website that normally does not interact with a smart contract, ex. most Web2 sites, can be hacked and start initiating transactions.

# Proposed Solution

Websites that interact with smart contracts should publish a list that let's wallets and other interfaces know what contracts the website is expected to call.

Any transaction from a site that requests a contract not on the list can then be treated as suspicious, and the user can be warned.

Also, as standards become more widely adopted, sites should be treated as unexpected to call a smart contract unless they explicitly declare it by providing a list.

# Technical Details

> ## Format

Reviewer: I was expecting a JSON schema with an explanation of each field here

Author: done

```jsx
//.well-known/contracts
{
  contracts: [
    {
      chainId: "string",
      contracted: "string",
    },
  ];
}
```

## Cases

### No well-known is present

> The site is not expected to call any smart contacts. If it does call a smart contract treat it as suspicious. This is how CORS works where by default all External requests are blocked.

Reviewer: link to docs on CORS

> ```jsx
> //.well-known/contracts
> undefined;
> ```
> 

Reviewer: ```suggestion
```

### wild card

Any smart contract is allowed from any chain

```jsx
//.well-known/contracts
{
  contracts: [
    {
      chainId: "*",
      contracted: "*",
    },
  ];
}
```

> ### Wild Card for Networks/Chains

Reviewer: ```suggestion
### Wildcard specifier for Networks/Chains
```

> In this sample only networks on Ethereum (1) and Polygon (137) are allowed.

Reviewer: Maybe include a link somewhere of where to find these IDs for those curious about it

Author: good idea, added

```jsx
//.well-known/contracts
{
	contracts:
	[
		{
			"chainId": "1",
			"contracted": "*"
		}
		{
			"chainId": "137",
			"contracted": "*"
		}
	]
}
```

> You could also set up TestNet IDs instead in your test environment.

Reviewer: Known testnet IDs could just be hard-coded into the wallets themselves. Similar to how Chrome ships with a list of sites to use HSTS on: https://hstspreload.org/ and for root CAs: https://www.chromium.org/Home/chromium-security/root-ca-policy/

Author: yeh most wallets already keep their own lists of networks. The main point here was to make sure a staging / test version of a website does not accidentally call contracts on a mainnet. Will see how to reword this better.

> ### Specific list

Reviewer: This should probably the first or second example and should include some sample JSON.

Check if a contract and chain are on the list. Any call not on the list should be blocked or at least warned about.

# Wallet implementation

> We recommend warning users. Outright blocking similar to CSP can be done once the standard is more widely accepted.

Reviewer: Anything future looking is usually discounted in proposals like these without concrete roadmaps. I'd leave this out or be more specific about what you mean by warning users. What types of contracts might a user sign? Where is the right place to warn? Who should be notified?

## Early Stage

The contract list is fully checked

> Sites without a contract list are not warned about since the standard is still being adopted.

Reviewer: Chrome does a good job of warning developers that they are missing some security feature and that it will break their sites in a coming version of the browser. I think it might make sense to warn the dev somehow that their contract is not set up right. Maybe the wallet should log something to the console or warn if using a testnet

## End Game

> Once the standard has been adopted allowing website developers with contracts to have time to at least add a wildcard on their site.

Reviewer: Sentence fragment

Any site that asks to sign a transaction for a contract without a contract list present should trigger a warning to the user.

> Contract list is fully checked

Reviewer: idk what this means. I'd just get rid of it

> 

Reviewer: Consider adding a section called "Alternatives Considered" and document what projects are currently doing

# Open Questions

## .well-known/contracts page vs Using Headers vs <meta> Tag

> Is the best way to share and expose this data through a page server on the domain, something like [mysite.com/.well-known/contracts](http://mysite.com/.well-known/contracts) , or by using headers on the site itself similar to CSP.

Reviewer: ```suggestion
Is the best way to expose this data via a [well-known URL](https://www.rfc-editor.org/rfc/rfc8615), ex. [mysite.com/.well-known/contracts](http://mysite.com/.well-known/contracts), or by using HTTP headers on web server that serves the site (similar to CSP).
```

> Another alternative is using the meta tag, similar to CSP [https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP)

Reviewer: The meta tag approach isn't really an alternative. It's more a by-product of the `http-equiv` attribute, which allows a web page to tell a user agent to treat it's content as if it is an HTTP header. You can't logically support the <meta> tag approach without the HTTP header approach since you'll break the semantics

Author: nice got it

# Website well-known Contracts [Draft]

### Draft

<aside>
🚧 This document is a work in progress, we are gathering feedback and suggestions to arrive at a complete specification.

</aside>

> 

Reviewer: ```suggestion

### Abstract

This proposal defines a new ".well-known" web resource (as defined in [RFC-8614](https://www.rfc-editor.org/rfc/rfc8615) for EVM smart contracts. It provides an interface for website authors to hint to crypto wallets what transactions (if any) to expect from a site during a user agent's web browsing session. In doing so, it helps crypto wallets better warn users about unexpected or potentially malicious transactions.

```

### Inspiration

Inspired by Content Security Policy (CSP) [https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP)

# Problem Definition

Currently a site that requests signatures can ask you to sign any transaction, even transactions that the site was not intended for.

For example, an NFT marketplace can request to sign a multi sig that drains your DAO Treasury, or a Defi app meant only to swap tokens can request to move all your NFTs out of your wallet.

## Code Injection

Malicious code injected into a website can access all the contents and permissions of a user's wallet, where the only thing protecting a user is their own decision making.

## Rogue Developers

One rogue developer at any dapp has the potential to drain all of the dapp users' wallets.

Code reviews can help with this, but there need to be clearer mechanisms to detect and prevent these attacks.

# Proposed Solution

> Websites that interact with smart contracts should publish a list that lets wallets and other interfaces know what contracts the website is expected to call.

Reviewer: ```suggestion
Websites that interact with smart contracts should publish a list at `/.well-known/contracts` that lets wallets and other interfaces know what contracts the website is expected to call.
```

Any transaction initiated on a site using a contract not on the list is treated as suspicious, and the user can be shown a warning

# Technical Details

## Outline

```jsx
//.well-known/contracts
{
  contracts: [
    {
      chainId: "string",
      contractId: "string",
    },
  ];
}
```

## Schema

```jsx
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "type": "object",
  "properties": {
    "contracts": {
      "type": "array",
      "items": [
        {
          "type": "object",
          "properties": {
            "chainId": {
              "type": "string"
            },
            "contractId": {
              "type": "string"
            }
          },
          "required": [
            "chainId",
            "contractId"
          ]
        }
      ]
    }
  },
  "required": [
    "contracts"
  ]
}
```

# Examples

## Specific list

Check if a contract and chain are on the list. Any call not on the list should be blocked or at least warned about.

> ```jsx

Reviewer: This should be json. And the keys should have quotes around them

```jsx
//.well-known/contracts
{
    contracts: [],
}
```

## No well-known is present

Don't do any checks. Similar to wildcard.

# Wallet implementation

## Warn the User

Show a warning to the user when triggering a transaction if the contract does not match the list published by the website.

This should be done for all transactions coming from the site that have a contract included.

## Log an Error

A console error should be logged to make it easier for the site's developers to detect when their site is having a problem due to transactions not matching their contract list.

## Sample Implementation

A sample library of making these checks will soon be released.

# Open Questions

## .well-known/contracts page vs Using Headers vs <meta> Tag

Is the best way to expose this data via a [well-known URL](https://www.rfc-editor.org/rfc/rfc8615), ex. [mysite.com/.well-known/contracts](http://mysite.com/.well-known/contracts), or by using HTTP headers on the web server that serves the site similar to [CSP](https://developer.mozilla.org/en-US/docs/Web/HTTP/CSP).

# FAQ and Decisions

## Contract Details, Functions, and CallData

> We chose to keep the spec to just the networkId and contractId to make it easier to agree on the core concept of the list and make the first version easier to implement and adopt.

Reviewer: ```suggestion
We chose to limit the scope of the spec to just the `networkId` and `contractId` to make it easier to agree on the core concept of the list and make the first version easier to implement and adopt.
```

> Future versions may have more details on the each contract, such as which function calls are expected, similar to the [Yearn Allowlist](https://medium.com/iearn/yearn-allowlist-71757d4e3cf4).

Reviewer: ```suggestion
Future revisions may have more details on each contract, such as which function calls are expected, similar to the [Yearn Allowlist](https://medium.com/iearn/yearn-allowlist-71757d4e3cf4).
```
